expression/builtin_time.go

```diff
@@ -332,83 +339,297 @@ type timeDiffFunctionClass struct {
+		default:
+			sig = &builtinTimeStringTimeDiffSig{baseDurationBuiltinFunc{bf}}
+		}
+	default:
+		switch arg1Tp {
+		case tpDuration:
+			sig = &builtinStringDurationTimeDiffSig{baseDurationBuiltinFunc{bf}}
+		case tpDatetime, tpTimestamp:
+			sig = &builtinStringTimeTimeDiffSig{baseDurationBuiltinFunc{bf}}
+		default:
+			sig = &builtinStringStringTimeDiffSig{baseDurationBuiltinFunc{bf}}
+		}
+	}
 	return sig.setSelf(sig), nil
 }
 
-type builtinTimeDiffSig struct {
-	baseBuiltinFunc
+type builtinDurationDurationTimeDiffSig struct {
+	baseDurationBuiltinFunc
 }
 
-func (b *builtinTimeDiffSig) getStrFsp(strArg string, fsp int) int {
-	if n := strings.IndexByte(strArg, '.'); n >= 0 {
-		lenStrFsp := len(strArg[n+1:])
-		if lenStrFsp <= types.MaxFsp {
-			fsp = int(math.Max(float64(lenStrFsp), float64(fsp)))
-		}
+// evalDuration evals a builtinDurationDurationTimeDiffSig.
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_timediff
+func (b *builtinDurationDurationTimeDiffSig) evalDuration(row []types.Datum) (d types.Duration, isNull bool, err error) {
+	sc := b.getCtx().GetSessionVars().StmtCtx
+	val0, isNull, err := b.args[0].EvalDuration(row, sc)
+	if isNull || err != nil {
+		return d, isNull, errors.Trace(err)
 	}
-	return fsp
+
+	val1, isNull, err := b.args[1].EvalDuration(row, sc)
+	if isNull || err != nil {
+		return d, isNull, errors.Trace(err)
+	}
+
+	d, err = val0.Sub(val1)
+	if err != nil {
+		return d, false, errors.Trace(err)
+	}
+
+	d.Duration, err = types.TruncateOverflowMySQLTime(d.Duration)
+	if types.ErrTruncatedWrongVal.Equal(err) {
+		err = sc.HandleTruncate(err)
+	}
+	return d, false, errors.Trace(err)
 }
 
-func (b *builtinTimeDiffSig) convertArgToTime(sc *variable.StatementContext, arg types.Datum, fsp int) (t types.Time, err error) {
-	// Fix issue #3923, see https://github.com/pingcap/tidb/issues/3923,
-	// TIMEDIFF() returns expr1 − expr2 expressed as a Duration value. expr1 and expr2 are Duration or date-and-time expressions,
-	// but both must be of the same type. if expr is a string, we first try to convert it to Duration, if it failed,
-	// we then try to convert it to Datetime
-	switch arg.Kind() {
-	case types.KindString, types.KindBytes:
-		strArg := arg.GetString()
-		fsp = b.getStrFsp(strArg, fsp)
-		t, err = types.StrToDuration(sc, strArg, fsp)
-	case types.KindMysqlDuration:
-		t, err = arg.GetMysqlDuration().ConvertToTime(mysql.TypeDuration)
-	default:
-		t, err = convertDatumToTime(sc, arg)
+type builtinTimeTimeTimeDiffSig struct {
+	baseDurationBuiltinFunc
+}
+
+// evalDuration evals a builtinTimeTimeTimeDiffSig.
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_timediff
+func (b *builtinTimeTimeTimeDiffSig) evalDuration(row []types.Datum) (d types.Duration, isNull bool, err error) {
+	sc := b.getCtx().GetSessionVars().StmtCtx
+	val0, isNull, err := b.args[0].EvalTime(row, sc)
```

Comment:
how about change `val0` to `lhs`, `val1` to `rhs` ?

Reply:
lhs means `left-hand size`? I think it is not suit in here.

Reply:
it is widely used in two-parameter functions